python,stdlib: Improve exit event return

BobbyRBruce opened the pull request:
[WIP: Not ready for review].

BobbyRBruce committed:
stdlib: Add 'hypercall' exit events

Harshil2107 committed:
misc: add hypercall event to m5ops

Harshil2107 committed:
sim: add m5 hypercall pseudo instruction

Harshil2107 committed:
sim: add hypercall functions to pybind

Harshil2107 committed:
sim: udpate simulator.py to use hypercalls

Harshil2107 committed:
misc: add example to test hyper call

- this commit also adds hypercall to old style exit events for testing

Harshil2107 committed:
misc: add hypercall to scons

Harshil2107 committed:
sim: Made hypercall work with gem5

- now we can define behaviour per specific number returned from m5 hypercall

Harshil2107 commented:
@powerjg @BobbyRBruce : I got the new hyper call events to work with this last commit. I will clean up the files and push the final version but I tested this by replacing the m5 exit in the disk images to `m5 hypercall` followed by 1,2 and 3 for the three exits we have.

We dont need to define a handler explicitly anymore.

Harshil2107 committed:
stdlib,sim: Cleanup Hypercall implementation

BobbyRBruce committed:
tests: Mv test_event_event.py -> test_exit_event.py

This was a typo.

BobbyRBruce committed:
stdlib,tests: Add hypercall exit event tests

Note the inclusion of the binary and it's source. This needs moved to
gem5-resources.

BobbyRBruce committed:
stdlib: Tidy up the Exit event revamp

erin-le committed:
sim, util: trigger hypercall with signal from external process

This commit adds a utility script for sending a signal and
optionally a payload to a running gem5 simulation. It also adds
a signal handler to gem5 to receive the signal and process the
payload.

erin-le committed:
sim, util: Change how shared memory is used in external hypercall

This commit changes how shared memory is used when a hypercall
is triggered by an external process. Unlinking the shared
memory has been moved to the transmitter side, and the signal
handler in gem5 writes a "done" message to the shared memory to
let the transmitter know to unlink it. The shared memory name
has also been changed to include the gem5 process's PID, so
separate gem5 processes won't have the same shared memory.

powerjg commented on src/python/gem5/simulate/exit_handler.py:
```diff
@@ -0,0 +1,207 @@
+    @overrides(ExitHandler)
+    def _process(self, simulator: "Simulator") -> None:
+        print("Second exit: Started `after_boot.sh` script")
+        print("Switching to Timing CPU")
```
This line is not generic and should be removed.

powerjg commented on src/python/gem5/simulate/exit_handler.py:
```diff
@@ -0,0 +1,207 @@
+class KernelBootedExitHandler(ExitHandler):
+    @overrides(ExitHandler)
+    def _process(self, simulator: "Simulator") -> None:
+        print("First exit: kernel booted")
```
"First" and "Second" (below) should probably be removed. We can't guarantee that there won't be some new event added before this.

Also, we should probably not use "print" but some kind of logging. Maybe this is a change for another patch, though.

powerjg commented on src/python/gem5/simulate/simulator.py:
```diff
@@ -91,6 +97,32 @@ class Simulator:
         "compatible with the gem5 standard library.",
     }
 
+    def get_default_exit_handler_id_map(cls) -> Dict[int, Type[ExitEvent]]:
+        default_map = {
+            # The default exit handlers for the older "generated-based" exit
+            # events.
+            0: ClassicGeneratorExitHandler,
+            # The default exit handler for when the the kernel in a FS
+            # simulation has completed booting.
+            1: KernelBootedExitHandler,
+            # The default exit handler for when the `after_boot.sh` script has
+            # completed execution in FS simulations which utilize this.
+            2: AfterBootExitHandler,
+            # The default exit handler triggered when the post-boot script has
+            # completed execution in FS simulation. Note: it is common for this
+            # to be the the defacto end of the simulation as the post-boot
+            # script is used to execute the user's workload.
+            3: AfterBootScriptExitHandler,
+            # The default exit handler for scheduled exit event, such as those
+            # scheduled by the user via `scheduleTickExitAbsolute` or
+            # `scheduleTickExitFromCurrent`.
+            4: ScheduledExitEventHandler,
+        }
+        assert all(
+            i >= 0 for i in default_map.keys()
+        ), "Exit handler mapped to ID <0 in default map."
+        return default_map
```
I still think this can automatically be generated (and extended) without having to modify the stdlib code (and rebuild gem5). I can mock this up on Thursday.

BobbyRBruce replied:
This can be extended and modified without rebuilding gem5. This is just the default allocation but it can be extended/modified via the configuration script as needed..

powerjg reviewed (commented):
A couple of small things from a quick skim of the first few files. Also, *many files are missing copyright*.

BobbyRBruce committed:
sim,util: External Hypercall signal 'SIGRTMIN' -> 'SIGHUP'

MacOS does not support real-time signals like 'SIGRTMIN'. Using this
makes "src/sim/init_signals.cc" uncompilable on Mac systems and would
make use of the external hypercall transmitter impossible for Mac users.

Given 'SIGUSR1' and 'SIGUSR2' are already used in gem5, this patch
changes the signal to 'SIGHUP'; a signal typically used to prompt
system daeomons to reload their configuration files. As this doesn't
apply to gem5, it seems save to use in this manner.

powerjg commented on src/sim/init_signals.hh:
```diff
@@ -40,6 +42,9 @@ void initSignals();
 // separate out sigint handler so that we can restore the python one
 void initSigInt();
+void initSigRtmin();
+std::string extractStringFromJSON(std::string& full_str, std::string start_str,
```
Is this really a full "json" parser or more restricted? I think it's more restricted with just {'key':'value'}.

Assuming I'm right, you should change the name to make that more clear. Also, adding some comments in the function to make the restriction clear is important as well.

powerjg commented on src/sim/sim_events.cc:
```diff
@@ -54,19 +54,38 @@ namespace gem5
 
 GlobalSimLoopExitEvent::GlobalSimLoopExitEvent(Tick when,
                                                const std::string &_cause,
-                                               int c, Tick r)
+                                               int c, Tick r,
+                                               uint64_t hypercall_id,
+        std::map<std::string, std::string> payload)
```
Why not create a new type for hypercalls that doesn't have the extraneous parameters. I don't think we'll ever need the code, repeat, or cause parameters for the hypercall. IMO, we should specialize more and reuse old code less.

In fact, I feel like this could be the other way around where we use the hypercall for the "old" way with the payload being `{'cause': <whatever>}`

powerjg commented on src/sim/sim_exit.hh:
```diff
@@ -55,6 +55,21 @@ void exitSimLoop(const std::string &message, int exit_code = 0,
 void exitSimLoopNow(const std::string &message, int exit_code = 0,
                     Tick repeat = 0, bool serialize = false);
 
+void exitSimLoopWithHypercall(const std::string &message, int exit_code,
```
As said above, I don't think we need these extra parameters here. But, let me know if I'm mistaken.

powerjg commented on tests/gem5/stdlib/configs/hypercall-exit-workload/include/gem5/asm/generic/m5ops.h:
These files shouldn't be here. Just use the files in gem5/include

powerjg commented on util/hypercall_external_signal/transmitter.py:
copyright...

powerjg commented on util/m5/src/command/hypercall.cc:
copyright

powerjg reviewed (commented):
I don't like the default dict because it separates the definition of the class from the hypercall number. I mocked up a way to improve this. I haven't tested it, but I don't know why it wouldn't work. This will also essentially automatically override the exit handler if you specify your own implementation. If you want to use one handler for multiple different exit events, I'm sure we could figure out how to do that. Let me know what you think.

https://github.com/powerjg/gem5/commit/73b3ae63438c32fbe8dc175c337c7b84fd2f7c6c

BobbyRBruce committed:
tests,stdlib: Cleanup Hypercall tests

BobbyRBruce committed:
stdlib: Add missing copyright notice

BobbyRBruce committed:
stdlib: Remove testing/placeholder processes for ExitHandler

BobbyRBruce committed:
stdlib: Add ExitHandler for WorkBegin and WorkEnd hypercalls

BobbyRBruce committed:
stdlib,configs: Fix exit handling example config

BobbyRBruce commented:
Note on this: As there is an urgency to have _a_ version of this available for work at UC Davis I have merged it prematurely. The comments about things to be improved here are still valid and will be addressed

BobbyRBruce commented:
This PR has been cherry-picked to the v24.1.1.0 staging branch: https://github.com/gem5/gem5/tree/release-staging-v24-1-1-0
